# !/usr/bin/env python
PACKAGE = "relocalization"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("debug_en", bool_t, 0, "Debug enable", False)
gen.add("pub_prior_pcd_en", bool_t, 0, "Publish prior pcd enable", False)
gen.add("map_frame", str_t, 0, "Map frame", "map")
gen.add("odom_frame", str_t, 0, "Odom frame", "odom")
gen.add("base_frame", str_t, 0, "Base frame", "base_link")
gen.add("lidar_frame", str_t, 0, "Lidar frame", "livox_frame")
gen.add("pcd_in_topic", str_t, 0, "Cloud topic for submap", "/cloud_registered")
gen.add("imu_in_topic", str_t, 0, "Imu topic for align gravity", "/livox/imu")
gen.add("initialpose_in_topic", str_t, 0, "Initialpose topic", "/initialpose")

gen.add("num_threads", int_t, 0, "The threads for gicp registration", 4)
gen.add("num_neighbors", int_t, 0, "The neighbor nums of point", 10)
gen.add("max_dist_sq", double_t, 0, "The distance between tow point", 1.0)
gen.add("max_iterations", int_t, 0, "The iteration times", 30)
gen.add("align_time_threshold", int_t, 0, "The time threshold for valid registration", 100)
gen.add("global_leaf_size", double_t, 0, "The resolution for global map of gicp", 0.25)
gen.add("registered_leaf_size", double_t, 0, "The resolution for scan", 0.1)

gen.add("x", double_t, 0, "The transform x for global map", 0.0)
gen.add("y", double_t, 0, "The transform y for global map", 0.0)
gen.add("z", double_t, 0, "The transform z for global map", 0.0)
gen.add("yaw", double_t, 0, "The transform yaw for global map", 0.0)
gen.add("roll", double_t, 0, "The transform roll for global map", 0.0)
gen.add("pitch", double_t, 0, "The transform pitch for global map", 0.0)

gen.add("registration_frequency", double_t, 0, "The frequency for registration", 4)
gen.add("pose_update_frequency", double_t, 0, "The frequency for broadcaster transform", 10)
gen.add("transform_tolerance", double_t, 0, "The broadcaster time tolerance", 0.1)

gen.add("max_level", int_t, 0, "", 6)
gen.add("min_level_resolution", double_t, 0, "", 0.5)
gen.add("map_leaf_size", double_t, 0, "", 0.2)
gen.add("source_leaf_size", double_t, 0, "", 0.1)
gen.add("min_scan_range", double_t, 0, "", 0.0)
gen.add("max_scan_range", double_t, 0, "", 100.0)
gen.add("score_percentage_threshold", double_t, 0, "", 0.9)
gen.add("timeout_sec", int_t, 0, "", 0)
gen.add("yaw_min_range", double_t, 0, "", 0.0)
gen.add("yaw_max_range", double_t, 0, "", 0.0)
gen.add("roll_min_range", double_t, 0, "", 0.0)
gen.add("roll_max_range", double_t, 0, "", 0.0)
gen.add("pitch_min_range", double_t, 0, "", 0.0)
gen.add("pitch_max_range", double_t, 0, "", 0.0)

gen.add("restore_defaults", bool_t, 0, "Retsore the default configuration", False)

exit(gen.generate(PACKAGE, "relocalization", "RelocalizationCfg"))